pkg/coveragedb/spanner.go

```diff
@@ -248,3 +250,71 @@ func NsDataMerged(ctx context.Context, projectID, ns string) ([]TimePeriod, []in
 	}
 	return periods, totalRows, nil
 }
+
+// DeleteGarbage checks for the dangling session data.
+// And deletes the dangling records in batches.
+// The batch size is 10k rows. Spanner limit is 80k mutations.
+// DeleteGarbage returns the amount of rows deleted.
+// In case of the error, some rows may be deleted but not counted.
+func DeleteGarbage(ctx context.Context) (int64, error) {
+	batchSize := 10_000
+	client, err := NewClient(ctx, os.Getenv("GOOGLE_CLOUD_PROJECT"))
+	if err != nil {
+		return 0, fmt.Errorf("coveragedb.NewClient: %w", err)
+	}
+	defer client.Close()
+
+	iter := client.Single().Query(ctx, spanner.Statement{
+		SQL: `SELECT session, filepath
+					FROM files
+					WHERE NOT EXISTS (
+						SELECT 1
+						FROM merge_history
+						WHERE merge_history.session = files.session
+					)`})
+	defer iter.Stop()
+
+	var totalDeleted atomic.Int64
+	eg, _ := errgroup.WithContext(ctx)
+	var batch []spanner.Key
+	for {
+		row, err := iter.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return 0, fmt.Errorf("iter.Next: %w", err)
+		}
+		var r struct {
+			Session  string
+			Filepath string
+		}
+		if err = row.ToStruct(&r); err != nil {
+			return 0, fmt.Errorf("row.ToStruct: %w", err)
+		}
+		batch = append(batch, spanner.Key{r.Session, r.Filepath})
+		if len(batch) > batchSize {
```

Comment:
Nit: if you want your batches to be specifically of batchSize size, you'd better `batch = append(batch, ...)` after this if block.

Reply:
I expect to eventually see more tables pointing this sessionID. It means every loop we'll add 1-2 more mutations. It is not important to split exactly by 10k.
Thanks for the corner cases verification!